x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/aggregation/blockhash/TimeSeriesBlockHash.java

```diff
@@ -0,0 +1,126 @@
+
+    private final int tsHashChannel;
+    private final int timestampChannel;
+    private final Rounding.Prepared preparedRounding;
+    private final DriverContext driverContext;
+    private final BytesRefHash tsidHashes;
+
+    long tsidOrdinal = -1;
+    BytesRef previousTsidHash;
+    long previousTimestamp;
+
+    public TimeSeriesBlockHash(int tsHashChannel, int timestampChannel, Rounding rounding, DriverContext driverContext) {
+        super(driverContext.blockFactory());
+        this.tsHashChannel = tsHashChannel;
+        this.timestampChannel = timestampChannel;
+        if (rounding != null) {
+            this.preparedRounding = rounding.prepareForUnknown();
+        } else {
+            this.preparedRounding = null;
+        }
+        this.driverContext = driverContext;
+        this.tsidHashes = new BytesRefHash(1, blockFactory.bigArrays());
+    }
+
+    @Override
+    public void close() {
+        Releasables.close(tsidHashes);
+    }
+
+    @Override
+    public void add(Page page, GroupingAggregatorFunction.AddInput addInput) {
+        BytesRefBlock tsHashBlock = page.getBlock(tsHashChannel);
+        BytesRefVector tsHashVector = Objects.requireNonNull(tsHashBlock.asVector());
+        try (var ordsBuilder = driverContext.blockFactory().newIntVectorBuilder(tsHashVector.getPositionCount())) {
+            LongBlock timestampIntervalBlock = page.getBlock(timestampChannel);
+            BytesRef spare = new BytesRef();
+            if (preparedRounding != null) {
+                for (int i = 0; i < tsHashVector.getPositionCount(); i++) {
+                    BytesRef tsHash = tsHashVector.getBytesRef(i, spare);
+                    long timestampInterval = preparedRounding.round(timestampIntervalBlock.getLong(i));
+                    if (tsHash.equals(previousTsidHash) == false || timestampInterval != previousTimestamp) {
+                        tsidOrdinal++;
+                        previousTsidHash = BytesRef.deepCopyOf(tsHash);
+                        previousTimestamp = timestampInterval;
+                    }
+                    ordsBuilder.appendInt(Math.toIntExact(tsidOrdinal));
+                }
+            } else {
+                for (int i = 0; i < tsHashVector.getPositionCount(); i++) {
+                    BytesRef tsHash = tsHashVector.getBytesRef(i, spare);
+                    if (tsHash.equals(previousTsidHash) == false) {
+                        tsidOrdinal++;
+                        previousTsidHash = BytesRef.deepCopyOf(tsHash);
+                        tsidHashes.add(previousTsidHash);
+                    }
+                    ordsBuilder.appendInt(Math.toIntExact(tsidOrdinal));
+                }
+            }
+            try (var ords = ordsBuilder.build()) {
+                addInput.add(0, ords);
+            }
+        }
+    }
+
+    @Override
+    public Block[] getKeys() {
+        int positions = (int) tsidHashes.size();
```

Comment:
I think the keys should consist of both the tsid and rounded timestamps.

Reply:
added via: 1ed5b904dabfc299fddde03415fef6de3612b1e0